kbs/src/plugins/implementations/pkcs11.rs

```diff
@@ -171,6 +186,87 @@ impl Pkcs11Backend {
             _ => bail!("Illegal HTTP method. Only supports `GET` and `POST`"),
         }
     }
+
+    async fn wrap_key_handle(&self, body: &[u8], method: &Method) -> Result<Vec<u8>> {
+        match *method {
+            Method::POST => self.wrapkey_wrap(body).await,
+            Method::GET => self.wrapkey_unwrap(body).await,
+            _ => bail!("invalid method"),
+        }
+    }
+
+    fn wrap_key_new(session: &mut Session, label: &Uuid) -> Result<()> {
+        let public_template = vec![
+            Attribute::Token(true),
+            Attribute::Private(false),
+            Attribute::KeyType(KeyType::RSA),
+            Attribute::Class(ObjectClass::PUBLIC_KEY),
+            Attribute::ModulusBits(4096.into()),
+            Attribute::Label(format!("{}-public", label).into()),
+        ];
+
+        let private_template = vec![
+            Attribute::Token(true),
+            Attribute::Private(true),
+            Attribute::KeyType(KeyType::RSA),
+            Attribute::Class(ObjectClass::PRIVATE_KEY),
+            Attribute::Label(format!("{}-private", label).into()),
+        ];
+
+        let (_, _) = session
+            .generate_key_pair(
+                &Mechanism::RsaPkcsKeyPairGen,
+                &public_template,
+                &private_template,
+            )
+            .context("unable to generate RSA wrap key pair")?;
+
+        Ok(())
+    }
+
+    async fn wrapkey_wrap(&self, body: &[u8]) -> Result<Vec<u8>> {
+        let pubkey_template = vec![Attribute::Label(
+            format!("{}-public", self.wrapkey_id).into(),
+        )];
+
+        let mut pubkey = self
+            .session
+            .lock()
+            .await
+            .find_objects(&pubkey_template)
+            .context("unable to find public wrap key in PKCS11 module")?;
+
+        let encrypted = self
+            .session
+            .lock()
+            .await
+            .encrypt(&Mechanism::RsaPkcs, pubkey.remove(0), body)
+            .context("unable to decrypt HTTP body with private wrap key")?;
```

Comment:
Is the message a copy/paste typo?

Reply:
Yes, good catch. Fixed in latest revision.